qt 6.0.0 (new formula)

jonaski opened the pull request.

carlocab commented:
Why not update the existing qt formula? Also, why does this have so many more dependencies than the qt5 formula?

jonaski commented:
- Qt 5 programs are not necessarily compatible with Qt 6. Qt 5 is LTS and will still be supported until 2023.
- I added the dependencies for building Qt 6. I can't speak for Qt 5, I didn't create that formula.

carlocab commented:
I think this formula would need to be the one named `qt`. The old `qt` formula should be renamed `qt@5`, and any formulae that are not compatible with qt6 should have their dependency switched to `qt@5`.

chenrui333 commented:
Yeah, I agree with @carlocab, we have examples like node and golang all have different versioned formulae.

jonaski commented:
I strongly disagree, but if that's the way things are done around here I will change it. But it will cause problems for all projects that use Qt 5 from homebrew in their CI. Qt 5 should not have been called just "qt" in the first place.

carlocab commented:
It's the way things are done here. See also `lua`, `openjdk`, etc.

SMillerDev commented:
> Qt 5 should not have been called just "qt" in the first place.

Homebrew policy is to have the latest version of any formula without a version in the name with the exceptions of openssl and python for purely maintenance reasons. If people using Homebrew QT in CI are not aware of the homebrew rules regarding versioning I'm not sure we can fix that for them.

chenrui333 commented:
> I strongly disagree, but if that's the way things are done around here I will change it. But it will cause problems for all projects that use Qt 5 from homebrew in their CI. Qt 5 should not have been called just "qt" in the first place.

the versioned formulae is just the way how homebrew managed the different versions of the *same* software, one way to mitigate the downstream impact is to add aliases.

jonaski commented:
I didn't realize that there were aliases, or I actually remember seeing that before now, but had forgotten. It's not a problem now since users can just use the qt5 or qt6 alias in their CI, even though I would have preferred it the other way.

carlocab commented:
Only one formula-file changed per commit, please. This means you'll need to split this up into two commits: one for `qt` + its aliases and another for `qt@5` + its aliases.

SMillerDev commented on Formula/qt.rb:
```diff
@@ -3,71 +3,77 @@
 
-  head "https://code.qt.io/qt/qt5.git", branch: "dev", shallow: false
-
-  livecheck do
-    url :head
-    regex(/^v?(\d+(?:\.\d+)+)$/i)
-  end
-
-  bottle do
-    cellar :any
-    sha256 "ac22ab5828d894518e42f00e254f1e36d5be4e5f3f1c08b3cd49b57819daaf2d" => :big_sur
-    sha256 "51ab78a99ff3498a236d15d9bed92962ddd2499c4020356469f7ab1090cf6825" => :catalina
-    sha256 "25c4a693c787860b090685ac5cbeea18128d4d6361eed5b1bfed1b16ff6e4494" => :mojave
-  end
```
This shouldn't be removed

SMillerDev commented on Formula/qt@5.rb:
```diff
@@ -0,0 +1,120 @@
+  sha256 "3a530d1b243b5dec00bc54937455471aaa3e56849d2593edb8ded07228202240"
+  license all_of: ["GFDL-1.3-only", "GPL-2.0-only", "GPL-3.0-only", "LGPL-2.1-only", "LGPL-3.0-only"]
+
+  bottle do
+    cellar :any
+    sha256 "ac22ab5828d894518e42f00e254f1e36d5be4e5f3f1c08b3cd49b57819daaf2d" => :big_sur
+    sha256 "51ab78a99ff3498a236d15d9bed92962ddd2499c4020356469f7ab1090cf6825" => :catalina
+    sha256 "25c4a693c787860b090685ac5cbeea18128d4d6361eed5b1bfed1b16ff6e4494" => :mojave
+  end
```
This shouldn't be here (yet)

jonaski commented:
- Qt 5 failed to build now because it picks the system ninja for webengine, adding `-no-feature-webengine-system-ninja` to configure
- Qt 6 fails to build in the CI because it includes the wrong math header file, however it builds fine for me locally. I had this program with the test locally, but I fixed that by unsetting CPATH, unsure how to do that with ninja via `cmake --build .`

jonaski commented:
- Qt 6.0.0 has a bug with GNU make on macOS, let's add `-cmake-generator 'Ninja'` to make sure it always uses Ninja.

jonaski commented:
@SMillerDev 
Most likely reason why qtwebengine for Qt 5 fails to build now is because of the @ in the path, see:
https://bugreports.qt.io/browse/QTBUG-79711?jql=text%20~%20%22macos%20ninja%22
How can we solve that?

jonaski commented:
Qt 6 build is still failing: 
```

2020-12-25T13:47:49.6573850Z In file included from /usr/local/Cellar/qt/6.0.0/lib/QtCore.framework/Headers/qmath.h:59:
2020-12-25T13:47:49.6576430Z /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/cmath:318:9: error: no member named 'signbit' in the global namespace
2020-12-25T13:47:49.6577610Z using ::signbit;
2020-12-25T13:47:49.6577970Z       ~~^
2020-12-25T13:47:49.6580000Z /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/cmath:319:9: error: no member named 'fpclassify' in the global namespace
2020-12-25T13:47:49.6581150Z using ::fpclassify;
2020-12-25T13:47:49.6581500Z       ~~^
2020-12-25T13:47:49.6583190Z /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/cmath:320:9: error: no member named 'isfinite' in the global namespace; did you mean 'finite'?
2020-12-25T13:47:49.6584360Z using ::isfinite;
2020-12-25T13:47:49.6584690Z       ~~^
2020-12-25T13:47:49.6586330Z /Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/usr/include/math.h:749:12: note: 'finite' declared here
```

jonaski commented:
Switched Qt 6 to build with GNU Make instead adding `CPATH=` to make. Have to disable QTuioTouchPlugin because of this bug: https://bugreports.qt.io/browse/QTBUG-84342
Qt 5 build fails with the same problem now with the cmath header stuff. Adding CPATH there too.

carlocab commented:
qt5 built just fine previously: https://github.com/Homebrew/homebrew-core/commit/945c55f8a6cc83a7e8ccc26b7d54502c13557bde

Why are these changes needed now?

jonaski commented:
I don't know, but it also fails with the same error in https://github.com/Homebrew/homebrew-core/pull/67170

carlocab commented:
I may be mistaken, but I believe those are errors on the ARM runner. Here, you're getting errors even on the Intel Big Sur runner. Perhaps there are changes you've made to the existing `qt` formula before renaming it to `qt@5` that are causing build failures?

jonaski commented:
I don't think so. Compare the files and you'll see what I've changed.

- I removed `-proprietary-codecs`, which was an invalid option.
- I changed `keg_only "Qt 5 has CMake issues when linked"` to `keg_only :versioned_formula`
- Removed head, livecheck and bottle.
- Had to add `-skip qtwebengine` because of https://bugreports.qt.io/browse/QTBUG-79711

carlocab commented:
>Had to add `-skip qtwebengine`

This seems to be a substantial change. I'm actually a little surprised the `@` makes it into `buildpath`; I always thought that only depended on the name of the tarball. Though, it seems weird that adding this should cause other build errors, so it may be unrelated. Very weird.

jonaski commented:
The cmath include build errors are in QtCore, It has something to do with including the wrong header file.
It has nothing to do with not building qtwebengine anymore.

fxcoudert commented:
I've killed the test, sorry, but `-skip qtwebengine` is simply not possible. I am trying to merge a fix to the current Qt version at https://github.com/Homebrew/homebrew-core/pull/67170 but we need to understand the failure so we can fix it, not work around it.

jonaski commented:
Everything builds now, but the test fails, I can't reproduce this locally. Does anyone know what's wrong here?

```
==> make CPATH=
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++ -c -pipe -stdlib=libc++ -O2 -std=gnu++1z  -arch x86_64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.1.sdk -mmacosx-version-min=10.14 -Wall -Wextra -fPIC -DQT_NO_DEBUG -DQT_CORE_LIB -I. -I/opt/homebrew/Cellar/qt/6.0.0/lib/QtCore.framework/Headers -I. -I/opt/homebrew/Cellar/qt/6.0.0/mkspecs/macx-clang -F/opt/homebrew/Cellar/qt/6.0.0/lib -o main.o main.cpp
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++ -stdlib=libc++ -headerpad_max_install_names  -arch x86_64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.1.sdk -mmacosx-version-min=10.14 -Wl,-rpath,@executable_path/../Frameworks -o hello main.o   -F/opt/homebrew/Cellar/qt/6.0.0/lib -framework QtCore -framework DiskArbitration -framework IOKit   
ld: warning: ignoring file /opt/homebrew/Cellar/qt/6.0.0/lib/QtCore.framework/QtCore, building for macOS-x86_64 but attempting to link with file built for macOS-arm64
Undefined symbols for architecture x86_64:
  "QArrayData::deallocate(QArrayData*, long long, long long)", referenced from:
      _main in main.o
      _main.cold.1 in main.o
  "QTextStream::operator<<(QString const&)", referenced from:
      _main in main.o
  "QTextStream::operator<<(char)", referenced from:
      _main in main.o
  "QCoreApplication::QCoreApplication(int&, char**, int)", referenced from:
      _main in main.o
  "QCoreApplication::~QCoreApplication()", referenced from:
      _main in main.o
  "QDebug::~QDebug()", referenced from:
      _main in main.o
  "QString::fromUtf8(QByteArrayView)", referenced from:
/opt/homebrew/Library/Homebrew/shims/scm/git --version
      _main in main.o
  "QMessageLogger::debug() const", referenced from:
      _main in main.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [hello] Error 1
```

fxcoudert commented:
- On Big Sur Intel, you can see `qt@5` fails to build with `ERROR: Unknown command line option '-proprietary-codecs'`
- On Big Sur Intel, qt (6) builds and tests fine. But most of the dependents are broken, because they need to be rebuilt
- On ARM, qt@5 fails to build with `ERROR: No value supplied to command line option 'skip'.`
- On ARM, qt (6) builds fine, but its test is failing because it's passing `-arch x86_64` to the compiler.

Is Apple Silicon supported on Qt 6.0.0?

fxcoudert commented:
@jonaski can you comment on what you changed? Force-push is the right thing to do, but the downside is that it makes it difficult to follow changes.

And also (if you can) about the question above: Is Apple Silicon supported on Qt 6.0.0?

jonaski commented:
For Qt 5 I removed `-proprietary-codecs` and `-skip` to fix compile.
I changed the test for Qt 6 to use cmake instead of qmake.
Apple silicon is scheduled for Qt 6.1: https://bugreports.qt.io/browse/QTBUG-85279
So not really supported I suppose.

qtwebengine builds fine if I rename qt@5 to qt5. The versioned package introduced the compile error. It's because qtwebengine
uses ninja which does not handle @ in paths, the problem is well-known and has been reported here: https://github.com/ninja-build/ninja/issues/1606
I have no fix for that, and I suggest to drop the versioned package and simply call it: qt5
I'll await feedback on how you want to do that.

fxcoudert commented:
> For Qt 5 I removed -proprietary-codecs and -skip to fix compile

We currently ship and build with these options, so I don't really understand why they're failing on this PR. What is the cause?

For the `@` issue, let's try to find a fix in brew: https://github.com/Homebrew/brew/pull/10162

jonaski commented:
`-proprietary-codecs` is webengine specific, and is invalid when webengine is disabled.
The extra -skip argument was a bug I introduced because I moved `-skip qtwebengine` to args.

jonaski commented:
I re-enabled qtwebengine now

jonaski commented:
The test for Qt 6 built with cmake fails in the CI now. It works fine for me locally.
Any idea what's wrong?

```
CMake Error in CMakeLists.txt:
  Imported target "Qt6::Core" includes non-existent path

    "/usr/local/mkspecs/macx-clang"

/usr/local/Homebrew/Library/Homebrew/shims/scm/git --version
  in its INTERFACE_INCLUDE_DIRECTORIES.  Possible reasons include:

  * The path was deleted, renamed, or moved to another location.

  * An install or uninstall procedure did not complete successfully.

  * The installation package was faulty and references files it does not
  provide.



CMake Error in CMakeLists.txt:
  Imported target "Qt6::Core" includes non-existent path

    "/usr/local/mkspecs/macx-clang"

  in its INTERFACE_INCLUDE_DIRECTORIES.  Possible reasons include:

  * The path was deleted, renamed, or moved to another location.

  * An install or uninstall procedure did not complete successfully.

  * The installation package was faulty and references files it does not
  provide.



CMake Error in CMakeLists.txt:
  Imported target "Qt6::Core" includes non-existent path

    "/usr/local/mkspecs/macx-clang"

  in its INTERFACE_INCLUDE_DIRECTORIES.  Possible reasons include:

  * The path was deleted, renamed, or moved to another location.

  * An install or uninstall procedure did not complete successfully.

  * The installation package was faulty and references files it does not
  provide.



-- Generating done
CMake Generate step failed.  Build files cannot be regenerated correctly.
/usr/bin/curl --version
/usr/libexec/java_home --xml --failfast
An exception occurred within a child process:
Error: qt: failed
  BuildError: Failed executing: cmake /private/tmp/qt-test-20201229-16402-kdr78g
```

fxcoudert commented:
Isn't it linked to the removal of this line?

> `keg_only "Qt 5 has CMake issues when linked"`

jonaski commented:
Added it back. Let's see if it builds now.
qtwebengine worked now. So thanks for the fix.

SMillerDev commented:
```
 In file included from qtbase/include/QtQuick/6.0.0/QtQuick/private/qquicktext_p.h:1:
qtquicktimeline/src/imports/timeline/../../../../../../../../../private/tmp/qt-20201229-35192-1r23nmg/qt-everywhere-src-6.0.0/qtdeclarative/src/quick/items/qquicktext_p.h:54:10: fatal error: 'qquickimplicitsizeitem_p.h' file not found
#include "qquickimplicitsizeitem_p.h"
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

SMillerDev commented on Formula/qt.rb:
```diff
@@ -3,10 +3,10 @@
-  sha256 "3a530d1b243b5dec00bc54937455471aaa3e56849d2593edb8ded07228202240"
+  url "https://download.qt.io/official_releases/qt/6.0/6.0.0/single/qt-everywhere-src-6.0.0.tar.xz"
+  mirror "https://mirrors.ocf.berkeley.edu/qt/official_releases/qt/6.0/6.0.0/single/qt-everywhere-src-6.0.0.tar.xz"
+  mirror "http://qt.mirrors.tds.net/qt/official_releases/qt/6.0/6.0.0/single/qt-everywhere-src-6.0.0.tar.xz"
```
Why are we downgrading to a plain HTTP mirror?

SMillerDev commented on Formula/qt.rb:
```diff
@@ -24,68 +24,75 @@ class Qt < Formula
   end
 
-  keg_only "Qt 5 has CMake issues when linked"
+  keg_only "Qt 6 has CMake issues when linked"
```
```suggestion
  keg_only "Qt #{version.major} has CMake issues when linked"
```

SMillerDev commented:
Lots of dependency tests failed, I expect that the linkage failures below will mostly just need a revision bump using `brew bump-revision --message "for qt" <name>` or will need to be updated to point to qt@5.
```
Error: 95 failed steps!
brew linkage --test audacious
brew test --retry --verbose audacious
brew linkage --test carla
brew test --retry --verbose carla
brew test --retry --verbose cgal
brew linkage --test codequery
brew linkage --test color-code
brew test --retry --verbose color-code
brew linkage --test djview4
brew linkage --test dspdfviewer
brew test --retry --verbose dspdfviewer
brew test --retry --verbose dynare
brew linkage --test gammaray
brew linkage --test gdal
brew test --retry --verbose gdal
brew linkage --test gdcm
brew linkage --test geant4
brew test --retry --verbose geant4
brew test --retry --verbose git-cola
brew test --retry --verbose gmt
brew test --retry --verbose gmt@5
brew linkage --test gnuplot
brew test --retry --verbose gnuplot
brew linkage --test gnuradio
brew linkage --test gpsbabel
brew test --retry --verbose gpsbabel
brew linkage --test icemon
brew test --retry --verbose icemon
brew linkage --test inspectrum
brew test --retry --verbose inspectrum
brew linkage --test kde-karchive
brew test --retry --verbose kde-karchive
brew linkage --test kde-kdoctools
brew test --retry --verbose kde-kdoctools
brew linkage --test kde-ki18n
brew test --retry --verbose kde-ki18n
brew linkage --test kde-threadweaver
brew test --retry --verbose kde-threadweaver
brew linkage --test libgr
brew test --retry --verbose mapserver
brew test --retry --verbose matplotplusplus
brew linkage --test mgba
brew test --retry --verbose mgba
brew linkage --test mlt
brew linkage --test octave
brew test --retry --verbose octave
brew linkage --test pc6001vx
brew test --retry --verbose pcl
brew test --retry --verbose pdal
brew test --retry --verbose points2grid
brew linkage --test poppler
brew linkage --test pushpin
brew test --retry --verbose pushpin
brew linkage --test pyqt
brew test --retry --verbose pyqt
brew linkage --test pyside
brew test --retry --verbose pyside
brew linkage --test qbs
brew test --retry --verbose qbs
brew linkage --test qca
brew test --retry --verbose qca
brew linkage --test qcachegrind
brew linkage --test qcli
brew test --retry --verbose qcli
brew linkage --test qjackctl
brew test --retry --verbose qjackctl
brew linkage --test qjson
brew test --retry --verbose qjson
brew linkage --test qmmp
brew test --retry --verbose qmmp
brew linkage --test qscintilla2
brew test --retry --verbose qscintilla2
brew linkage --test qsoas
brew test --retry --verbose qsoas
brew linkage --test qtads
brew linkage --test qtkeychain
brew test --retry --verbose qtkeychain
brew linkage --test quazip
brew test --retry --verbose quazip
brew linkage --test qwt
brew test --retry --verbose qwt
brew linkage --test qwtpolar
brew test --retry --verbose qwtpolar
brew linkage --test qxmpp
brew test --retry --verbose qxmpp
brew test --retry --verbose simple-tiles
brew linkage --test treefrog
brew test --retry --verbose treefrog
brew test --retry --verbose trimage
brew test --retry --verbose urh
brew linkage --test vtk
brew test --retry --verbose vtk
brew linkage --test xpdf
brew linkage --test zurl
brew test --retry --verbose zurl
```

paperchalice commented:
Hi, I also write some formulae for Qt6, I hope these formulae can help you.
[my version](https://github.com/Homebrew/homebrew-core/compare/master...paperchalice:qt)

fxcoudert commented:
@jonaski any specific reason you're closing this?
